src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticStartAnalysisScope.cs

```diff
@@ -617,14 +617,7 @@ public void RegisterOperationAction(DiagnosticAnalyzer analyzer, Action<Operatio
 
         protected AnalyzerActions GetOrCreateAnalyzerActions(DiagnosticAnalyzer analyzer)
         {
-            AnalyzerActions actions;
-            if (!_analyzerActions.TryGetValue(analyzer, out actions))
-            {
-                actions = new AnalyzerActions();
-                _analyzerActions[analyzer] = actions;
-            }
-
-            return actions;
+            return _analyzerActions.GetOrAdd(analyzer, _ => new AnalyzerActions());
```

Comment:
Investigated this further and following are some notes:

1. The entire machinery to register actions on scope or get registered actions has been non-thread safe since the driver was originally written by John Hamby few years ago.
2. Registration action callbacks have worked fine as we create a separate analysis context and [HostAnalysisScope(http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticStartAnalysisScope.cs,472) per analyzer, per callback. We could have still run into a race condition where a concurrent analyzer captured the `AnalysisContext` or `CompilationStartAnalysisContext` and invoked register action APIs from each nested actions, but it is unlikely for an analyzer to attempt such dynamic registration.
3. [GetAnalyzerActionsAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerManager.cs,146) invocations have worked fine as we only used to invoke them only once at end of compilation start actions in the core analyzer driver, and during telemetry callbacks from the IDE driver, which did not happen concurrently for a single analyzer.
4. With the introduction of `SymbolStartAction`, dynamic registration of actions can happen, but it will normally happen only on the `SymbolStartAnalysisContext` created per-symbol, per-analyzer, and hence making race conditions during registration still unlikely. However, the core analyzer driver now invokes the [GetAnalyzerActionsAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalyzerManager.cs,146) concurrently for each symbol, and this makes the race condition during this call extremely likely - note that this method is on the callstack on the bug being fixed here.

My recommendations:
1. +1 on the change in this PR, as it fixes the more likely race condition observed in the FailFast.
2. File a follow up issue to make `HostAnalysisScope` and its subtypes thread-safe, so we avoid all possible race conditions.

Reply:
➡️ Filed as a follow-up issue https://github.com/dotnet/roslyn/issues/30458